### PR TITLE
Upgrade codebase to polkadot-v0.9.23

### LINKED ISSUE #715: What is required to make Moonbeam's staking pallet allow locked tokens to be staked?
I'm opening this issue so we can track the investigation of Moonbeam's staking pallet in order to evaluate the work required to make it work with locked tokens (like the Parity staking pallet does). 

I'll report back and update this issue with some additional information after taking a look at the pallet.

### LINKED ISSUE #809: Upgrade codebase to polkadot-v0.9.23
**Dependencies**
Cumulus: https://github.com/paritytech/cumulus/tree/polkadot-v0.9.23
Moonbeam: https://github.com/PureStake/moonbeam/tree/v0.25.0
Nimbus: https://github.com/PureStake/nimbus/tree/moonbeam-polkadot-v0.9.23
ORML: https://github.com/open-web3-stack/open-runtime-module-library/tree/release-v0.9.23
Polkadot: https://github.com/paritytech/polkadot/tree/release-v0.9.23
Substrate: https://github.com/paritytech/substrate/tree/polkadot-v0.9.23

**Changes**
Cumulus: https://github.com/paritytech/cumulus/compare/polkadot-v0.9.19...polkadot-v0.9.23
Moonbeam: https://github.com/PureStake/moonbeam/compare/78db06c0203f61b35059304f7194ed5c10dcfda8...v0.25.0
Nimbus: https://github.com/PureStake/nimbus/compare/moonbeam-polkadot-v0.9.19...moonbeam-polkadot-v0.9.23
ORML: https://github.com/open-web3-stack/open-runtime-module-library/compare/polkadot-v0.9.19...polkadot-v0.9.23
Polkadot: https://github.com/paritytech/polkadot/compare/release-v0.9.19...release-v0.9.23
Substrate: https://github.com/paritytech/substrate/compare/polkadot-v0.9.19...polkadot-v0.9.23
Substrate-Parachain: https://github.com/substrate-developer-hub/substrate-parachain-template/compare/polkadot-v0.9.19...polkadot-v0.9.23

### PR DESCRIPTION
**Pull request by sea212**
Closes #809
Closes #715 

This PR updates the dependencies.

**Additional features**
- Use original repositories instead of forks (previously many forks from Moonbeam were used)
- Use references in `get_spot_price` and corresponding RPCs
- Implement Moonbeam's VRFs (generate randomness) in the client, add it to `author_rotateKeys` rpc and to `pallet-author-mapping`
- Build benchmarks from parachain modules only when parachain modules are activated by another feature
- Use `jsonrpsee` instead of `jsonrpc_core`
- Use patched Purestake libraries to ensure all modules are fetched from original repositories and make `try-runtime` work for `pallet-author-mapping`

**Notes**
Executing `try-runtime` will show decoding errors for `pallet-author-mapping`. This happens after `pre_upgrade` and `on_runtime_upgrade` and can be ignored - the storage is correctly migrated, as the `try-runtime` checks prove. In addition, the runtime upgrade was tested on a Battery Station clone. Unfortunately nodes have to upgrade their clients to be able to determine the author of the first block after the runtime upgrade.

### CODE REVIEW

File: zrml/swaps/rpc/src/lib.rs [hunk 1]
```diff
@@ -87,14 +92,15 @@ pub enum Error {
...
+impl From<Error> for i32 {
+    fn from(e: Error) -> i32 {
         match e {
-            Error::RuntimeError => 69,
+            Error::RuntimeError => 1,
         }
     }
 }
 
+/*
```
Comment (vivekvpandya) [inline zrml/swaps/rpc/src/lib.rs:103]: Please add some details in comment, for why this is commented.

File: node/src/cli.rs [hunk 2]
```diff
@@ -196,6 +199,16 @@ pub struct Cli {
...
 
+    /// Disable automatic hardware benchmarks.
+    ///
+    /// By default these benchmarks are automatically ran at startup and measure
+    /// the CPU speed, the memory bandwidth and the disk speed.
+    ///
+    /// The results are then printed out in the logs, and also sent as part of
+    /// telemetry, if telemetry is enabled.
+    #[clap(long)]
+    pub no_hardware_benchmarks: bool,
```
Comment (vivekvpandya) [inline node/src/cli.rs:210]: should we use name `hardware_benchmarks` ?

File: runtime/common/src/lib.rs [hunk 3]
```diff
@@ -51,6 +51,23 @@ macro_rules! decl_common_types {
...
 
+        #[cfg(feature = "parachain")]
+        pub type Executive = frame_executive::Executive<
+            Runtime,
+            Block,
+            frame_system::ChainContext<Runtime>,
+            Runtime,
+            AllPalletsWithSystem,
+            (
+                pallet_author_mapping::migrations::AddKeysToRegistrationInfo<Runtime>,
```
Comment (vivekvpandya) [inline runtime/common/src/lib.rs:62]: @sea212 could you please provide some context about these new migrations?

File: scripts/benchmarks/configuration.sh [hunk 4]
```diff
@@ -6,15 +6,15 @@ EXTERNAL_WEIGHTS_PATH="./runtime/common/src/weights/"
 export FRAME_PALLETS=( frame_system pallet_balances pallet_bounties pallet_democracy \
                 pallet_identity pallet_membership  pallet_multisig pallet_preimage \
                 pallet_proxy pallet_scheduler pallet_timestamp pallet_treasury \
-                pallet_utility pallet_vesting \
-                ) # pallet_collective pallet_grandpa )
+                pallet_utility pallet_vesting pallet_collective ) # pallet_grandpa )
```
Comment (vivekvpandya) [inline scripts/benchmarks/configuration.sh:9]: why pallet_collective was earlier disabled?

File: node/src/cli.rs (hunk 2 above)
Comment (sea212) [reply node/src/cli.rs:210]: Hardware benchmarks should be enabled by default and
```rust
pub no_hardware_benchmarks: bool
```
results in the client having a flag `--no-hardware-benchmarks` which will disable them. Thus I am in favor of keeping it like that.

File: scripts/benchmarks/configuration.sh (hunk 4 above)
Comment (sea212) [reply scripts/benchmarks/configuration.sh:9]: The benchmarks for `pallet_collective` failed before, now they work. See https://github.com/zeitgeistpm/zeitgeist/issues/822.

File: primitives/Cargo.toml [hunk 5]
```diff
@@ -1,15 +1,18 @@
...
-orml-tokens = { branch = "2022-06-v0.9.19", default-features = false, git = "https://github.com/zeitgeistpm/open-runtime-module-library" }
-orml-traits = { branch = "2022-06-v0.9.19", default-features = false, git = "https://github.com/zeitgeistpm/open-runtime-module-library" }
-parity-scale-codec = { default-features = false, features = ["derive", "max-encoded-len"], version = "3.0.0" }
+frame-support = { branch = "polkadot-v0.9.23", default-features = false, git = "https://github.com/paritytech/substrate" }
+frame-system = { branch = "polkadot-v0.9.23", default-features = false, git = "https://github.com/paritytech/substrate" }
+orml-currencies = { branch = "polkadot-v0.9.23", default-features = false, git = "https://github.com/open-web3-stack/open-runtime-module-library" }
+orml-tokens = { branch = "polkadot-v0.9.23", default-features = false, git = "https://github.com/open-web3-stack/open-runtime-module-library" }
+orml-traits = { branch = "polkadot-v0.9.23", default-features = false, git = "https://github.com/open-web3-stack/open-runtime-module-library" }
+parity-scale-codec = { default-features = false, features = [
+    "derive",
```
Comment (vivekvpandya) [inline primitives/Cargo.toml:9]: can you please update to taplo 0.7.0 and then run format script? New version avoids this kind of formatting.

File: scripts/benchmarks/configuration.sh (hunk 4 above)
Comment (vivekvpandya) [reply scripts/benchmarks/configuration.sh:9]: okay but there is still some investigation required for #822 or it failure was due to parity's bug and it got fixed?

File: runtime/common/src/lib.rs (hunk 3 above)
Comment (sea212) [reply runtime/common/src/lib.rs:62]: [`AddKeysToRegistrationInfo` adds a session key to author mapping](https://github.com/PureStake/moonbeam/pull/1407). This session key is used for Verifiable Random Functions (VRF) that can provide an improved source of randomness.
[`AddAccountIdToNimbusLookup` adds a reverse lookup](https://github.com/PureStake/moonbeam/pull/1525) `account_id` -> `nimbus_id` to `pallet_author_mapping` that is required.
[`SplitDelegatorStateIntoDelegationScheduledRequests`](https://github.com/PureStake/moonbeam/pull/1408) does plenty of things, please refer to the link.

Comment (vivekvpandya) [reply runtime/common/src/lib.rs:62]: :+1:

File: scripts/benchmarks/configuration.sh (hunk 4 above)
Comment (sea212) [reply scripts/benchmarks/configuration.sh:9]: It's failure was due to a bug in `pallet_collective` (uint bitwidth mismatch) and I am not sure if it got fixed, but we use a configuration that does not trigger that bug currently (we don't suffer any drawbacks from that either).

File: primitives/Cargo.toml (hunk 5 above)
Comment (sea212) [reply primitives/Cargo.toml:9]: I updated to taplo 0.7.0 and run the script, but it does not apply any changes. Do I have to revert all of that now and format again or is there another way?

Comment (vivekvpandya) [reply primitives/Cargo.toml:9]: unfortunately it seems so.

Comment (sea212) [reply primitives/Cargo.toml]: https://github.com/zeitgeistpm/zeitgeist/pull/811/commits/a350f05fd64af01cd41b38f9216f9ad20a3f0c51

**Review by vivekvpandya — DISMISSED** [review]

File: runtime/common/src/lib.rs [hunk 6]
```diff
@@ -1336,6 +1371,12 @@ macro_rules! create_runtime_api {
                     tx: <Block as BlockT>::Extrinsic,
                     block_hash: <Block as BlockT>::Hash,
                 ) -> TransactionValidity {
+                // Filtered calls should not enter the tx pool as they'll fail if inserted.
+                // If this call is not allowed, we return early.
+                if !<Runtime as frame_system::Config>::BaseCallFilter::contains(&tx.function) {
+                    return frame_support::pallet_prelude::InvalidTransaction::Call.into();
+                }
```
Comment (maltekliemann) [inline runtime/common/src/lib.rs:1378]:
```suggestion
                    // Filtered calls should not enter the tx pool as they'll fail if inserted.
                    // If this call is not allowed, we return early.
                    if !<Runtime as frame_system::Config>::BaseCallFilter::contains(&tx.function) {
                        return frame_support::pallet_prelude::InvalidTransaction::Call.into();
                    }
```

File: zrml/prediction-markets/fuzz/Cargo.toml [hunk 7]
```diff
@@ -6,9 +6,12 @@ test = false
...
 [dependencies]
 arbitrary = { features = ["derive"], version = "1.0" }
-frame-support = { branch = "moonbeam-polkadot-v0.9.19", default-features = false, git = "https://github.com/purestake/substrate" }
+frame-support = { branch = "polkadot-v0.9.23", default-features = false, git = "https://github.com/paritytech/substrate" }
 libfuzzer-sys = "0.4"
-zeitgeist-primitives = { features = ["arbitrary", "mock"], default-features = false, path = "../../../primitives" }
+zeitgeist-primitives = { features = [
+    "arbitrary",
+    "mock",
+], default-features = false, path = "../../../primitives" }
```
Comment (maltekliemann) [inline zrml/prediction-markets/fuzz/Cargo.toml:14]:
```suggestion
zeitgeist-primitives = { features = ["arbitrary", "mock"], default-features = false, path = "../../../primitives" }

File: zrml/prediction-markets/runtime-api/Cargo.toml [hunk 8]
```diff
@@ -1,6 +1,9 @@
 [dependencies]
-parity-scale-codec = { default-features = false, features = ["derive", "max-encoded-len"], version = "3.0.0" }
-sp-api = { branch = "moonbeam-polkadot-v0.9.19", default-features = false, git = "https://github.com/purestake/substrate" }
+parity-scale-codec = { default-features = false, features = [
+    "derive",
+    "max-encoded-len",
+], version = "3.0.0" }
```
Comment (maltekliemann) [inline zrml/prediction-markets/runtime-api/Cargo.toml:5]:
```suggestion
parity-scale-codec = { default-features = false, features = ["derive", "max-encoded-len"], version = "3.0.0" }
```

**Review by maltekliemann — APPROVED** [review]
Added a couple of formatting nitpicks, which I think should be fixed, but if you want to skip another CI cycle, I think we're good to go on this one. :thumbsup:

File: zrml/prediction-markets/fuzz/Cargo.toml (hunk 7 above)
Comment (sea212) [reply zrml/prediction-markets/fuzz/Cargo.toml]: https://github.com/zeitgeistpm/zeitgeist/pull/811/commits/3e31ec2675298fbd4b1273b8e0326e1604275247

**Review by maltekliemann — APPROVED** [review]
:thumbsup:

**Pull request merged**